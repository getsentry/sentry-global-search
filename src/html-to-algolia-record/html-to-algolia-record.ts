import { SearchHit } from '../sentry-global-search/lib/types';
import { hashObject } from './lib/hash-object';
import { getChildText } from './lib/get-child-text';
import { htmlToAST } from './lib/html-to-ast';
import { Meta } from './lib/types';
import { selectOne, selectAll, is } from 'css-select';

const INCLUDE = ['h1', 'h2', 'h3', 'p', 'li', '[data-index]', 'pre'];
const EXCLUDE = ['[data-noindex]'];

const isDescendant = (testNode, highNode) => {
  let test = testNode.parent;
  while (test) {
    if (highNode === test) return true;
    test = test.parent;
  }
  return false;
};

/**
 * Create algolia record objects from HTML. Intended for use with the rendered
 * HTML generated from Markdown, which has a reliably flat structure.  See the
 * README for further details about the strategy this uses.
 *
 * @param html The HTML string
 * @param meta Additional content to be included in the record. At a minimum
 *             must include `title` and `url`
 */
export const parseRecordsFromHTML = async (
  html: string,
  meta: Meta,
  baseSelector?: string
) => {
  const { title } = meta;
  const records: SearchHit[] = [];

  // This object is merged into each record and is mutated to track headings
  //and relevance
  const initialValues = {
    // id of the last heading seen
    anchor: '',
    // title of the page
    section: title,
    // h1 = 100, h2 = 90, h3 = 80
    // Relevance decreases as nesting increases
    sectionRank: 100,
    // Incremented for each record. Relevance decreases as position increases
    position: 0,
  };

  // Fetch the initial AST
  let dom = await htmlToAST(html);

  if (baseSelector) dom = selectOne(baseSelector, dom);

  dom = selectAll(`${INCLUDE.join(',')}`, dom);
  dom = dom.filter(x =>
    is(x, `:not(${EXCLUDE.map(x => `${x}, ${x} *`).join(',')})`)
  );

  dom.reduce((acc, el) => {
    const isChildOfExistingElement = !!dom.find(x => isDescendant(el, x));

    if (isChildOfExistingElement) return acc;
    const text = getChildText(el).trim();

    // Update the context when we get a heading.
    if (/h[1-3]/.test(el.name)) {
      const level = parseInt(el.name[1], 10);
      acc.anchor = el.attribs.id;
      acc.section = text;
      acc.sectionRank = 100 * (1 - 0.1 * (level - 1));

      // We don't want to create records for titles, we just want the data, so
      // we'll return here.
      return acc;
    }

    // Keep track of where in the doc this is. Lower stuff is less important
    acc.position++;

    const record = {
      text,
      ...meta,
      popularity: meta.popularity ?? Number.MAX_SAFE_INTEGER,
      ...acc,
    };
    records.push({ objectID: hashObject(record), ...record });

    return acc;
  }, initialValues);

  return records;
};
