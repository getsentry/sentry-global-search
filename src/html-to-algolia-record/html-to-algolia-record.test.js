import sampleDocument from './lib/sample-document';
import { parseRecordsFromHTML } from './html-to-algolia-record';

describe('HTML to Algolia record', () => {
  let document = {};

  beforeAll(async () => {
    document = await sampleDocument();
  });

  test('meets expectations', async () => {
    const { html, title, url } = document;
    const records = await parseRecordsFromHTML(html, { title, url });
    expect(records).toMatchSnapshot();
  });
});
