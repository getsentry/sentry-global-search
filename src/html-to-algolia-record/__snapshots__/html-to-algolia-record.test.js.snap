// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML to Algolia record meets expectations 1`] = `
Array [
  Object {
    "anchor": undefined,
    "objectID": "c2c149b6e273a4a57ac2ca4e49edad4c",
    "position": 1,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "For an overview of what Sentry does, take a look at the Sentry workflow.",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "2d4c55c8caf66611e2df1a7189f74b0e",
    "position": 2,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "Sentry is designed to be both simple to get off the ground and powerful to grow into. If you have never used Sentry before, this tutorial helps you get started.",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "42a9ce2c5af7ba13b3aa77a1d7a11405",
    "position": 3,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "Getting started with Sentry is a three step process:",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "d5999f9e12fb12f3b9433cbf021017df",
    "position": 4,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "Sign up for an account This will Test children",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "4995f334480a7270b36392938640433c",
    "position": 5,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "Install your SDK",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "46e1f32abd6f252266eca6b99cfcdd61",
    "position": 6,
    "section": "Getting started",
    "sectionRank": 100,
    "text": "Configure your SDK",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": "release",
    "objectID": "33577b321eac83aed35219875320252d",
    "position": 7,
    "section": "release",
    "sectionRank": 80,
    "text": "This content will fall under the \\"release\\" header",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "1a61d5f8ced6c6349c2d2b80022b8e2b",
    "position": 8,
    "section": "Install an SDK",
    "sectionRank": 90,
    "text": "Sentry captures data by using an SDK within your application’s runtime. These are platform specific and allow Sentry to have a deep understanding of how your application works.",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "128d1587fa22bb13f9578e87c5c32827",
    "position": 9,
    "section": "Install an SDK",
    "sectionRank": 90,
    "text": "Install our Python SDK using pip:",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "e40d8347761c9fb897eeb794e73584ac",
    "position": 10,
    "section": "Install an SDK",
    "sectionRank": 90,
    "text": "$ pip install --upgrade sentry-sdk==0.16.2",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "30c74743c6987e7ef26a3674dba4c0ac",
    "position": 11,
    "section": "Install an SDK",
    "sectionRank": 90,
    "text": "$ This will be indexed because it contains \`data-index\`",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "bec502430b9bfbd22ef28bcb93f135a3",
    "position": 12,
    "section": "Install an SDK",
    "sectionRank": 90,
    "text": "$ Pre blocks are indexed by default",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "427ec32e7d8c8d3a4ebf20a8319206e7",
    "position": 13,
    "section": "Configure the SDK",
    "sectionRank": 80,
    "text": "After you completed setting up a project in Sentry, you’ll be given a value which we call a DSN, or Data Source Name. It looks a lot like a standard URL, but it’s actually just a representation of the configuration required by the Sentry SDKs. It consists of a few pieces, including the protocol, public key, the server address, and the project identifier.",
    "title": "Getting started",
    "url": "https://example.com",
  },
  Object {
    "anchor": undefined,
    "objectID": "2610c839a4f8f96a2aa32f544ddd5299",
    "position": 14,
    "section": "Configure the SDK",
    "sectionRank": 80,
    "text": "This is text in an li and this text is a paragraph .",
    "title": "Getting started",
    "url": "https://example.com",
  },
]
`;
